core/src/components/textarea/textarea.tsx

```diff
@@ -705,55 +707,98 @@ Developers can use the "legacy" property to continue using the legacy form marku
+     * causes the slots and label to appear vertically offset from each
+     * other when the label isn't floating above the input. This doesn't
+     * apply to the outline fill, but this was not accounted for to keep
+     * things consistent.
+     */
+    const labelShouldFloat =
+      labelPlacement === 'stacked' || (labelPlacement === 'floating' && (hasValue || hasFocus || hasStartEndSlots));
+
     return (
       <Host
         class={createColorClasses(this.color, {
           [mode]: true,
-          'has-value': this.hasValue(),
-          'has-focus': this.hasFocus,
+          'has-value': hasValue,
+          'has-focus': hasFocus,
+          'label-floating': labelShouldFloat,
           [`textarea-fill-${fill}`]: fill !== undefined,
           [`textarea-shape-${shape}`]: shape !== undefined,
           [`textarea-label-placement-${labelPlacement}`]: true,
           'textarea-disabled': disabled,
         })}
       >
-        <label class="textarea-wrapper">
+        {/**
+         * htmlFor is needed so that clicking the label always focuses
+         * the textarea. Otherwise, if the start slot has something
+         * interactable, clicking the label would focus that instead
+         * since it comes before the textarea in the DOM.
+         */}
+        <label class="textarea-wrapper" htmlFor={inputId}>
           {this.renderLabelContainer()}
-          <div class="native-wrapper" ref={(el) => (this.textareaWrapper = el)}>
-            <textarea
-              class="native-textarea"
-              ref={(el) => (this.nativeInput = el)}
-              id={inputId}
-              disabled={disabled}
-              autoCapitalize={this.autocapitalize}
-              autoFocus={this.autofocus}
-              enterKeyHint={this.enterkeyhint}
-              inputMode={this.inputmode}
-              minLength={this.minlength}
-              maxLength={this.maxlength}
-              name={this.name}
-              placeholder={this.placeholder || ''}
-              readOnly={this.readonly}
-              required={this.required}
-              spellcheck={this.spellcheck}
-              cols={this.cols}
-              rows={this.rows}
-              wrap={this.wrap}
-              onInput={this.onInput}
-              onChange={this.onChange}
-              onBlur={this.onBlur}
-              onFocus={this.onFocus}
-              onKeyDown={this.onKeyDown}
-              {...this.inheritedAttributes}
-            >
-              {value}
-            </textarea>
+          <div class="textarea-wrapper-inner">
+            {/**
+             * Some elements have their own padding styles which may
+             * interfere with slot content alignment (such as icon-
+             * only buttons setting --padding-top=0). To avoid this,
+             * we wrap both the start and end slots in separate
+             * elements and apply our padding styles to that instead.
+             */}
+            <div class="start-slot-wrapper">
```

Comment:
If desired, I can hide these wrappers via CSS or avoid rendering them entirely if there isn't any slotted content provided. I figured it'd be better to keep things simple since keeping them doesn't hurt anything aside from having an extra node in the DOM, but I'm cool with making that change.